DM-38054: Add DataCoordinate attribute access to DimensionRecords.

TallJimbo opened the pull request:
## Checklist

- [x] ran Jenkins
- [x] added a release note for user-visible changes to `doc/changes`

TallJimbo committed:
Add DataCoordinate attribute access to DimensionRecords.

timj reviewed (approved):
Looks good to me. Can we also have an explicit test for the common usage? In test_dimensions.py there is a line:

```python
self.assertEqual(dataId.timespan, dataId.records["visit"].timespan)
```

can we add a line underneath it as:

```python
self.assertEqual(dataId.timespan, dataId.visit.timespan)
```
?

timj commented on python/lsst/daf/butler/core/dimensions/_coordinate.py:
```diff
@@ -1045,3 +1049,14 @@ def hasRecords(self) -> bool:
+        try:
+            return self._record(name)
+        except KeyError:
+            raise AttributeError(name) from None
```
It wouldn't hurt to also have a test on an unexpanded dataId to make sure it raises AttributeError? Then I imagine the follow up will be that people will want to distinguish between an attribute that is failing because it's not part of the dataId vs one that is failing because the records were not available (so maybe the error message should mention the distinction).

TallJimbo committed:
Override DataCoordinate pickle implementations.

Without this, the default pickle implementation apparently tries to
do getattr on instances before they're fully constructed while catching
the exceptions.  And that leads to infinite recursion if one writes
a __getattr__ that involves accessing attributes added at construction.

timj commented on python/lsst/daf/butler/core/dimensions/_coordinate.py:
```diff
@@ -156,6 +156,10 @@ class DataCoordinate(NamedKeyMapping[Dimension, DataIdValue]):
+
+    See also
+    --------
+    :ref:`lsst.daf.butler-dimensions_data_ids`
```
I tried building the docs but I didn't see this section included. May want to consider embedding it into the Notes section instead.

TallJimbo replied:
It's there (or at least it is for me) - it's a little call-out box _above_ the Notes section, unlike the way it appears in the source.

timj replied:
Ah. Yes. I see it now I know what I'm looking for. Sorry.
